File: algo_notation.py
"""Tunisian algorithm notation in the Algorithme column, whatever the model wrote.

The course writes `Q ← A div B`, `R ← A mod B`, `≠`, `≤`, `≥`, `ET`, `OU`,
`NON`, `Vrai`, `Faux`. Python writes `//`, `%`, `!=`, `<=`, `>=`, `and`, `or`,
`not`, `True`, `False`. The model knows both and sometimes lets a Python
operator slip into the Algorithme column (`Res ← X // Y`), which a student
copies into their exercise book and is marked wrong for - the one mistake
that most damages their trust in the tutor.

The prompts forbid it (prompts.py, rule 1); this module is the safety net for
when the model does it anyway. It rewrites the operators ONLY:

  - in the Algorithme column of an `Algorithme | Python` table (the Python
    column, the explanations and the student's own quoted lines are left
    alone - in a CODE answer, citing the student's `//` is the point);
  - outside double-quoted strings, so `Ecrire ("Remise de 5 %")` keeps its %.

The same rewrite runs in the browser (ui/src/lib/algoNotation.js) for the
streamed answer and for saved discussions; here it runs before the syntax
checker, so a line the student will see corrected is not reported as a
violation, and each correction is logged so the model's slips stay visible.
"""

from __future__ import annotations

import re

# Longest operators first: `<=` before `<-`, `//` before anything with `/`.
_OPERATORS: list[tuple[re.Pattern, str]] = [
    (re.compile(r"\s*//\s*"), " div "),
    (re.compile(r"\s*%\s*"), " mod "),
    (re.compile(r"\s*==\s*"), " = "),
    (re.compile(r"\s*!=\s*"), " ≠ "),
    (re.compile(r"\s*<=\s*"), " ≤ "),
    (re.compile(r"\s*>=\s*"), " ≥ "),
    (re.compile(r"\s*<-\s*"), " ← "),
    (re.compile(r"\band\b"), "ET"),
    (re.compile(r"\bor\b"), "OU"),
    (re.compile(r"\bnot\b"), "NON"),
    (re.compile(r"\bTrue\b"), "Vrai"),
    (re.compile(r"\bFalse\b"), "Faux"),
]
_INNER_SPACES = re.compile(r"(?<=\S) {2,}(?=\S)")


def _fix_code(segment: str) -> tuple[str, int]:
    count = 0
    for pattern, replacement in _OPERATORS:
        segment, n = pattern.subn(replacement, segment)
        count += n
    if count:
        leading = segment[: len(segment) - len(segment.lstrip(" "))]
        body = _INNER_SPACES.sub(" ", segment[len(leading) :])
        segment = leading + body
    return segment, count


def to_algo_notation(line: str) -> tuple[str, int]:
    """One Algorithme line in course notation, and how many operators changed.
    Text inside double quotes is copied as it is."""
    out: list[str] = []
    count = 0
    # Alternating code / "string" pieces; an unclosed quote runs to the end.
    for piece in re.split(r'("[^"]*"?)', line):
        if piece.startswith('"'):
            out.append(piece)
        else:
            fixed, n = _fix_code(piece)
            out.append(fixed)
            count += n
    # A rewrite next to a string can leave "div  (" or trailing spaces.
    result = "".join(out)
    return (result.rstrip() if count else result), count


def _cells(row: str) -> list[str]:
    inner = row.strip()
    if inner.startswith("|"):
        inner = inner[1:]
    if inner.endswith("|"):
        inner = inner[:-1]
    return inner.split("|")


def normalize_answer(markdown: str) -> tuple[str, int]:
    """The answer with every `Algorithme | Python` table's Algorithme column in
    course notation. Returns (text, number of operators rewritten)."""
    lines = markdown.split("\n")
    total = 0
    algo_index: int | None = None
    for i, line in enumerate(lines):
        stripped = line.strip()
        if not stripped.startswith("|"):
            algo_index = None  # the table ended
            continue
        cells = _cells(stripped)
        headers = [c.strip().lower() for c in cells]
        if "algorithme" in headers and "python" in headers:
            algo_index = headers.index("algorithme")
            continue
        if algo_index is None or algo_index >= len(cells):
            continue
        if re.fullmatch(r"\s*:?-{2,}:?\s*", cells[algo_index]):
            continue  # the |---|---| separator
        fixed, n = to_algo_notation(cells[algo_index])
        if n:
            original = cells[algo_index]
            pad_left = original[: len(original) - len(original.lstrip())] or " "
            cells[algo_index] = f"{pad_left}{fixed.strip()} "
            indent = line[: len(line) - len(line.lstrip())]
            lines[i] = f"{indent}|{'|'.join(cells)}|"
            total += n
    return "\n".join(lines), total

File: ui/src/lib/algoNotation.js
/**
 * Tunisian algorithm notation for one Algorithme-column line - the browser
 * twin of app/grading/algo_notation.py (same rules, same tests' cases).
 *
 * The course writes `div`, `mod`, `=`, `≠`, `≤`, `≥`, `←`, `ET`, `OU`, `NON`,
 * `Vrai`, `Faux`; Python writes `//`, `%`, `==`, `!=`, `<=`, `>=`, `<-`-style
 * arrows, `and`, `or`, `not`, `True`, `False`. The prompts forbid the Python
 * spellings in the Algorithme column, but when the model slips, a student
 * copying `Res ← X // Y` into their exercise book is marked wrong - so the
 * column is displayed (and copied) in course notation regardless.
 *
 * Text inside double quotes is never touched: `Ecrire ("Remise de 5 %")`.
 * Used by remarkAlgoTable.js, so it covers the streamed answer and every
 * saved discussion alike; the Python column and the prose are left as written.
 */

const OPERATORS = [
  [/\s*\/\/\s*/g, " div "],
  [/\s*%\s*/g, " mod "],
  [/\s*==\s*/g, " = "],
  [/\s*!=\s*/g, " ≠ "],
  [/\s*<=\s*/g, " ≤ "],
  [/\s*>=\s*/g, " ≥ "],
  [/\s*<-\s*/g, " ← "],
  [/\band\b/g, "ET"],
  [/\bor\b/g, "OU"],
  [/\bnot\b/g, "NON"],
  [/\bTrue\b/g, "Vrai"],
  [/\bFalse\b/g, "Faux"],
];

function fixCode(segment) {
  let count = 0;
  let out = segment;
  for (const [pattern, replacement] of OPERATORS) {
    out = out.replace(pattern, () => {
      count += 1;
      return replacement;
    });
  }
  if (!count) return [segment, 0];
  const leading = out.match(/^ */)[0];
  const body = out.slice(leading.length).replace(/(?<=\S) {2,}(?=\S)/g, " ");
  return [leading + body, count];
}

/** [line in course notation, number of operators rewritten]. */
export function toAlgoNotation(line) {
  let count = 0;
  const pieces = String(line).split(/("[^"]*"?)/);
  const out = pieces.map((piece) => {
    if (piece.startsWith('"')) return piece;
    const [fixed, n] = fixCode(piece);
    count += n;
    return fixed;
  });
  const result = out.join("");
  return [count ? result.trimEnd() : result, count];
}

/** Several lines (a cell with <br>s, or the whole column for copying). */
export const algoText = (text) =>
  String(text)
    .split("\n")
    .map((line) => toAlgoNotation(line)[0])
    .join("\n");
